diff --git a/core/res/res/values/config.xml b/core/res/res/values/config.xml
--- a/core/res/res/values/config.xml
+++ b/core/res/res/values/config.xml
@@ -8087,4 +8087,5 @@
 
     <string name="config_gnssPsdsType">qualcomm</string>
     <java-symbol type="string" name="config_gnssPsdsType" />
+    <bool name="config_setVibrationAmplitudeBeforeOn">false</bool>
 </resources>
diff --git a/core/res/res/values/symbols.xml b/core/res/res/values/symbols.xml
--- a/core/res/res/values/symbols.xml
+++ b/core/res/res/values/symbols.xml
@@ -6635,4 +6635,5 @@
   <java-symbol type="drawable" name="ic_cross_device_sync_airplane" />
   <!-- Disable the Video Capability over IWLAN when Wi-Fi Calling is turned off  -->
   <java-symbol type="bool" name="config_disable_video_capability_when_wfc_off" />
-  </resources>
+    <java-symbol type="bool" name="config_setVibrationAmplitudeBeforeOn" />
+</resources>
diff --git a/services/core/java/com/android/server/vibrator/VibrationSettings.java b/services/core/java/com/android/server/vibrator/VibrationSettings.java
--- a/services/core/java/com/android/server/vibrator/VibrationSettings.java
+++ b/services/core/java/com/android/server/vibrator/VibrationSettings.java
@@ -221,6 +221,11 @@
 
         // Update with current values from settings.
         update();
+    }
+
+    boolean shouldSetAmplitudeBeforeOn() {
+        return mContext.getResources().getBoolean(
+                com.android.internal.R.bool.config_setVibrationAmplitudeBeforeOn);
     }
 
     public void onSystemReady() {
diff --git a/services/core/java/com/android/server/vibrator/SetAmplitudeVibratorStep.java b/services/core/java/com/android/server/vibrator/SetAmplitudeVibratorStep.java
--- a/services/core/java/com/android/server/vibrator/SetAmplitudeVibratorStep.java
+++ b/services/core/java/com/android/server/vibrator/SetAmplitudeVibratorStep.java
@@ -141,6 +141,9 @@
                             + duration + "ms");
         }
         int stepId = conductor.nextVibratorCallbackStepId(getVibratorId());
+        if (conductor.vibrationSettings.shouldSetAmplitudeBeforeOn()) {
+            vibrator.setAmplitude(amplitude);
+        }
         long vibratorOnResult = vibrator.on(getVibration().id, stepId, duration);
         handleVibratorOnResult(vibratorOnResult);
         getVibration().stats.reportVibratorOn(vibratorOnResult);
diff --git a/services/core/java/com/android/server/vibrator/DeviceAdapter.java b/services/core/java/com/android/server/vibrator/DeviceAdapter.java
--- a/services/core/java/com/android/server/vibrator/DeviceAdapter.java
+++ b/services/core/java/com/android/server/vibrator/DeviceAdapter.java
@@ -46,6 +46,7 @@
      * VibratorInfo. They can be applied in a chain to a mutable list before a new VibrationEffect
      * instance is created with the final segment list.
      */
+    private final PrebakedFallbackAdapter mPrebakedFallbackAdapter;
     private final List<VibrationSegmentsAdapter> mSegmentAdapters;
     /**
      * The vibration segment validators that can validate VibrationEffectSegments entries based on
@@ -54,9 +55,8 @@
     private final List<VibrationSegmentsValidator> mSegmentsValidators;
 
     DeviceAdapter(VibrationSettings settings, SparseArray<HalVibrator> vibrators) {
+        mPrebakedFallbackAdapter = new PrebakedFallbackAdapter(settings.getFallbackEffects());
         mSegmentAdapters = Arrays.asList(
-                // Replace unsupported prebaked effects with fallback
-                new PrebakedFallbackAdapter(settings.getFallbackEffects()),
                 // Updates primitive delays to hardware supported pauses
                 new PrimitiveDelayAdapter(),
                 // Add extra ramp down segments as needed
@@ -79,6 +79,18 @@
         for (int i = 0; i < vibrators.size(); i++) {
             mAvailableVibratorIds[i] = vibrators.keyAt(i);
         }
+    }
+
+    private DeviceAdapter(DeviceAdapter original, SparseArray<VibrationEffect> fallbackEffects) {
+        mPrebakedFallbackAdapter = new PrebakedFallbackAdapter(fallbackEffects);
+        mSegmentAdapters = original.mSegmentAdapters;
+        mSegmentsValidators = original.mSegmentsValidators;
+        mAvailableVibrators = original.mAvailableVibrators;
+        mAvailableVibratorIds = original.mAvailableVibratorIds;
+    }
+
+    DeviceAdapter withFallbacks(SparseArray<VibrationEffect> fallbackEffects) {
+        return new DeviceAdapter(this, fallbackEffects);
     }
 
     SparseArray<HalVibrator> getAvailableVibrators() {
@@ -105,7 +117,8 @@
 
         VibratorInfo info = vibrator.getInfo();
         List<VibrationEffectSegment> newSegments = new ArrayList<>(composed.getSegments());
-        int newRepeatIndex = composed.getRepeatIndex();
+        int newRepeatIndex = mPrebakedFallbackAdapter.adaptToVibrator(
+                info, newSegments, composed.getRepeatIndex());
 
         int adapterCount = mSegmentAdapters.size();
         for (int i = 0; i < adapterCount; i++) {
diff --git a/services/core/java/com/android/server/vibrator/VibrationStepConductor.java b/services/core/java/com/android/server/vibrator/VibrationStepConductor.java
--- a/services/core/java/com/android/server/vibrator/VibrationStepConductor.java
+++ b/services/core/java/com/android/server/vibrator/VibrationStepConductor.java
@@ -20,6 +20,7 @@
 import android.annotation.Nullable;
 import android.os.Build;
 import android.os.VibrationEffect;
+import android.os.vibrator.Flags;
 import android.os.vibrator.PrebakedSegment;
 import android.os.vibrator.PrimitiveSegment;
 import android.os.vibrator.PwleSegment;
@@ -187,10 +188,16 @@
 
         waitForVibrationParamsIfRequired();
 
+        if (Flags.removeHidlSupport()) {
+            mVibration.fillFallbacks(vibrationSettings::getFallbackEffect);
+        }
+
         // Scale resolves the default amplitudes from the effect before scaling them.
         mVibration.scaleEffects(mVibrationScaler);
 
-        if (!mVibration.adaptToDevice(mDeviceAdapter)) {
+        DeviceAdapter deviceAdapter = Flags.removeHidlSupport()
+                ? mDeviceAdapter.withFallbacks(mVibration.mFallbacks) : mDeviceAdapter;
+        if (!mVibration.adaptToDevice(deviceAdapter)) {
             // Unable to adapt vibration effect for playback. This likely indicates the presence
             // of unsupported segments. The original effect will be ignored.
             return false;
