diff --git a/hal/Android.bp b/hal/Android.bp
index 75d06cf..60cf89c 100644
--- a/hal/Android.bp
+++ b/hal/Android.bp
@@ -75,6 +75,10 @@ target_board_platform_cc_library_shared {
         "true": ["-DSOUND_TRIGGER_ENABLED"],
         default: [],
     }) + select(soong_config_variable("ANDROID", "target_board_platform"), {
+        "sdm845": [
+            "-DPLATFORM_SDM845",
+            "-DMAX_TARGET_SPECIFIC_CHANNEL_CNT=2",
+        ],
         "kona": [
             "-DPLATFORM_KONA",
             "-DMAX_TARGET_SPECIFIC_CHANNEL_CNT=4",
